File: 74. Search a 2D Matrix.cpp
class Solution {
public:
    bool searchMatrix(const vector<vector<int>>& matrix, int target) {
        int n = matrix.size();
        int m = matrix[0].size();
        int left = 0, right = n * m;

        while (left < right) {
            int mid = left + (right - left) / 2;

            if (matrix[mid / m][mid % m] == target) return true;
            else if (matrix[mid / m][mid % m] < target) left = mid + 1;
            else right = mid;
        }

        return false;
    }
};

File: 74. Search a 2D Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int n = matrix.length;
        int m = matrix[0].length;
        int left = 0, right = n * m;

        while(left < right){
            int mid = left + (right - left) / 2;

            // System.out.println(mid + "-" + left + "-" + right + "=" + matrix[mid / m][mid % m] + " for " + target);
            if(matrix[mid / m][mid % m] == target) return true;
            else if(matrix[mid / m][mid % m] < target) left = mid + 1;
            else right = mid;
        }

        return false;
    }
}